### DISCUSSION TITLE
Disable directive that doesn't trigger Unused eslint-disable directive warning

### DISCUSSION THREAD

**Discussion by thomasballinger** [eslint/eslint#18867] [Ideas]
I have a CLI tool that generates .js and .d.ts files, similar to OpenAPI or GraphQL or other code generation. When I generate this code, for my users' convenience I'd like to avoid it having lint errors.

With no configuration on the user's part these files can dodge prettier with
```
/* prettier-ignore-start */
...
/* prettier-ignore-end */
```

However I can't fully dodge eslint v9 because now a
```
/* eslint-disable */
```
that is unused [causes a warning by default](https://github.com/eslint/eslint/pull/17879), and for some eslint configurations the generated files have no code issues.

Is there a way to opt out entirely of eslint based on file contents, say a directive that won't be warned about like this, for autogenerated files? Could one be created, or could `/* eslint-disable */` be special-cased not to trigger this warning?

One suggestion on Twitter was to add an issue that eslint would normally warn about but I'd rather not make triggering someone else's linter more likely. 

https://x.com/ballingt/status/1832214760097771804

**Comment by thomasballinger** [comment]
I'd like a way to say that no one should care about lints to a file because it is generated by a tool the person with access to the ESLInt output will never control. Tools like [TanStack Router also generate files](https://github.com/TanStack/router/blob/601adfbdcd961f68f7611da11db7469c1ad7f0f7/packages/router-generator/src/config.ts#L23-L28) that would like to opt out of being linted.

### Proposal 1
Simplest would be changing the behavior of `/* eslint-disable */` not to count as an unused disable directive. This isn't great for devs trying to use this warning to eliminate these though.

### Proposal 2
Maybe the default behavior of the "unused disable directives" warning could be changed not to include `/* eslint-disable */` but a separate, non-default stricter setting could be created that did. Say `reportUnusedDisableDirectives: 'strict'`.

### Proposal 3
What about a new disable directive like
```
/* eslint-disable-for-generated-file */
```
that it is not possible to make trigger an unused disable directive warning. Projects that generate code would have to add this new directive explicitly to their output.

**Reply by SeanCassiere** [reply]
I'd advocate for **_proposal 3_** here, mostly since it'd be a directive specifically allowing library authors to disable eslint for generated files.

Proposals 1 and 2 from above, don't seem too viable to me as it is clear that the eslint team made a conscious effort to have the `/* eslint-disable */` differ from its previous behaviour and having to add move levers into the eslint config file is an additional API-surface to worry about, for both the eslint team and the 3rd-party library authors (as they'd need to have to document the process of setting the correct value on the `reportUnusedDisableDirectives` property.

My only request here, is that the implementation of **proposal 3** (`/* eslint-disable-for-generated-file */`) would guarantee that eslint's static analysis would abort early all times. The 3rd-party library authors would be able inject something like this at the top of their generated files to allow for backward compatibility with eslint v8.
```ts
// my-generated-file.gen.ts
/* eslint-disable-for-generated-file */ <- abort early here

/* eslint-disable */ <- fallback for v8, as it wouldn't have the above directive.

// ...
```

**Reply by mdjermanovic** [reply]
> ### Proposal 3
> What about a new disable directive like
> 
> ```
> /* eslint-disable-for-generated-file */
> ```
> 
> that it is not possible to make trigger an unused disable directive warning. Projects that generate code would have to add this new directive explicitly to their output.

This was discussed in https://github.com/eslint/eslint/issues/18028 and https://github.com/eslint/rfcs/pull/118, and we eventually decided not to accept the proposal (please see the linked issue and the RFC for more details). The recommended solution is to add those files/directories to `ignores`.

**Reply by thomasballinger** [reply]
Ah yep looks like exactly this, with the only addition being that now ESLint 9 now has unused directive detection on by default.

It looks like that effort was abandoned after concerns were raise, I'll ask there about whether reviving it is appropriate.

**Comment by SeanCassiere** [comment]
Here from [tanstack/router](https://github.com/tanstack/router). We too do find it odd that the `/* eslint-disable */` directive would count towards an unused disable directive.

It cannot be avoided that the js-ecosystem has historically used generated files for outputting user-accessible APIs and/or types, with notable examples like Relay (and the entire graphql ecosystem in general), OpenAPI client generators, autogenerated react-hooks, and others.

The contents of these generated files are controlled by the authors of said libraries, with it being well known, that any user changes to these files are heavily discouraged and would often be overridden (either automatically or via a _package.json_ script).

A solution from the eslint team here would most certainly be helpful, as these files that are being generated, are not user code. They just happen live next to it.

**Comment by JoshuaKGoldberg** [comment]
> With no configuration on the user's part these files can dodge prettier with

Aside: I would be interested in seeing Prettier add its own 'unused directive' warning... If your code actually does match Prettier and the comment does nothing, then there might not be a need for the comment. I've seen this happen when folks work around a quirk in Prettier that later gets fixed.

https://github.com/prettier/prettier/issues/16737

**Comment by JoshuaKGoldberg** [comment]
Thinking just as an individual, not on behalf of any team: I don't totally follow the user story here. If some file is auto-generated, then why bother running ESLint with strict unused-directive checks on it? If users aren't meant to modify the file, then what's the point of letting them know about potential issues in the file? Only two edge cases come to mind...

Example one: are you trying to use formatting (`prettier --write`) + linting (`eslint --fix`) to standardize the file's style to the consuming project's? In that case it makes sense to remove any unnecessary disable directives for cluttering up the file. Maybe you could set unused directives to `warn` + be auto-fixable, and have `--max-warnings 1` or higher?

Example two: are you trying to validate that the file doesn't have any logical issues per your lint settings? For example, are you running ESLint on the file every time the package that generates the file is updated, to make sure the new package version doesn't introduce a regression?

**Reply by thomasballinger** [reply]
> If some file is auto-generated, then why bother running ESLint with strict unused-directive checks on it? If users aren't meant to modify the file, then what's the point of letting them know about potential issues in the file?

Ah sorry, I've not communicated this well. I don't want to let them know about issues! I want these files to be ignored by ESLint.

This is the case where files generated by a library or CLI tool like TanStack Router or Convex should not be linted, but I don't want to try to modify a eslint.config.js programmatically, or include additional instructions in these tools for how to modify an eslint.config.js file. I'd like to be able to include something in these files that causes formatting or lint issues not to be reported.

**Reply by thomasballinger** [reply]
I'd like to make these files "unlintable," regardless of configuration. To speed up linting the user can configure their linter to ignore whatever file pattern these generated files happen to follow, but for cases where there's no obvious wildcard pattern describing these files I'd like to be able specify this in the file itself. If the developer using my generated files wants to use strict unused-directive checks, great! I don't want to require them to turn this off to use these generated files.

**Reply by JoshuaKGoldberg** [reply]
Gotcha, thanks, that does make more sense. My gut instinct is that the right path would, unfortunately, be to either:

* Encourage putting these auto-generated files in automatically ignored locations, such as how Prisma puts some files in `node_modules/`
* Tell users to modify their ESLint/Prettier/etc. ignores to ignore the autogenerated files

ESLint could add in an `/* eslint-ignore-file */` directive or some such (that's what I'd call it). Then you'd have both ESLint and Prettier disabled for the file. But what about users on other linters, such as Biome, Oxlint, and TSSLint? They'd need these directives too, no? Eventually the files would look like:

```ts
/* biome-ignore-file */
/* deno-ignore-file */
/* eslint-ignore-file */
/* oxlint-ignore-file */
/* prettier-ignore-file */
/* tsslint-ignore-file */
```

Now suppose someone were to propose & get ratified a unified standard, like `/* ignore-file */` or something similar. But then what if _some_ users actually do want to run, say, only ESLint? But others want to run, say, only Prettier? I think this gets really hairy.

**Reply by thomasballinger** [reply]
That's indeed the direction this is going, e.g.
https://github.com/TanStack/router/blob/7ff37871e689acd3a0a24efae83801d65f5cf3b9/examples/react/start-basic/app/routeTree.gen.ts#L1-L7
starts with
```
/* prettier-ignore-start */

/* eslint-disable */

// @ts-nocheck

// noinspection JSUnusedGlobalSymbols
```

Yeah, until some generic `/* generated-file-please-ignore */` directive standard each tool either needs a directive or will have to be configured. Agree it's messy! This is coming up now because it's a strategy available for the default eslint 8 configuration but with eslint 9 default configuration is no longer available. Without this capability [this (good) change](https://github.com/eslint/eslint/pull/17879) is more disruptive.

Fair that ESLint 9 is a breaking change, projects like TanStack Router of course will need to do some work to update! But if this existed it would allow projects that generate code to keep doing things the old way.

**Reply by thomasballinger** [reply]
As a maintainer of one of these tools I'm trying to figure out how to proceed: for my users on ESLint 8 I'd like to keep these directives in their generated files so I don't introduce lint warnings, but to support users that are upgrading to ESLint 9 I'd like to remove these, or else they need to disable `reportUnusedDisableDirectives` and won't be able to use it. Any suggestions welcome!

**Reply by SeanCassiere** [reply]
> As a maintainer of one of these tools I'm trying to figure out how to proceed: for my users on ESLint 8 I'd like to keep these directives in their generated files so I don't introduce lint warnings, but to support users that are upgrading to ESLint 9 I'd like to remove these, or else they need to disable `reportUnusedDisableDirectives` and won't be able to use it. Any suggestions welcome!

+1 on this.

The fact that the reporting behaviour of `/* eslint-disable */` changed between v8 and v9 when using `reportUnusedDisableDirectives` is not ideal.

---

> * Encourage putting these auto-generated files in automatically ignored locations, such as how Prisma puts some files in node_modules/

This is not applicable for all libs and for some, this outright just wouldn't work (likely for functionality reasons). Prisma is more of an exception rather than the norm here, whereas all the graphql tools put the generated files in user-land (`src/`, `app/`).

---

> * Tell users to modify their ESLint/Prettier/etc. ignores to ignore the autogenerated files 

IMO, this is what we are trying to avoid.

Whilst I don't disagree that every library having to implement a `/* ignore this generated file */` directive isn't pretty, it's considerably better than the alternative of having to walk the user through editing the `xxx.config.js` file for their chosen linter and/or formatter.

It is undeniable that eslint and prettier are the biggest in this space. Prettier has an option for ignoring a file.
Eslint **_had_** an option in V8 that correctly disabled linting on the file when running with `reportUnusedDisableDirectives`. This behaviour was great! An actual escape hatch. Now, however, that escape hatch is no longer available 😞

Still at advocate for proposal 3 as mentioned here: https://github.com/eslint/eslint/discussions/18867#discussioncomment-10676403

**Reply by mdjermanovic** [reply]
> +1 on this.
> 
> The fact that the reporting behaviour of `/* eslint-disable */` changed between v8 and v9 when using `reportUnusedDisableDirectives` is not ideal.

The behavior of `/* eslint-disable */` is same in v8 and v9. Reporting unused disable directives is different: it's on by default in v9. But you can set it to `"off"` in the config file to get the v8 behavior.

https://eslint.org/docs/latest/use/configure/configuration-files#reporting-unused-disable-directives

```js
// eslint.config.js
export default [
    {
          linterOptions: {
            reportUnusedDisableDirectives: "off"
        }
    }
];
```

**Reply by jamrsim** [reply]
@mdjermanovic  answer sure seems like a workable solution to me.  Thank you!

**Reply by SeanCassiere** [reply]
> Tell users to modify their ESLint/Prettier/etc. ignores to ignore the autogenerated files

@jamrsim We wanted to avoid having users touch their eslint config at all. You'd be surprised at the number of people who don't want to touch their eslint config once they've gotten it to a happy state. As mentioned by @mdjermanovic, this was discussed in https://github.com/eslint/eslint/issues/18028 and https://github.com/eslint/eslint/issues/18028, but ultimately nothing came of it.

For the time being, we've taken on the task of saving a few threads where we've instructed users on the process of ignoring these files to use them as a resource for others who come in wondering why their eslint setup is indicating that the generated file has errors in them.
